1613.1 - Data File Summary model for storing status based on parsed errors

andrew-jameson opened the pull request:
## Summary of Changes
_Provide a brief summary of changes_
Pull request closes #_

## How to Test
_List the steps to test the PR_
These steps are generic, please adjust as necessary.
```
cd tdrs-frontend && docker-compose -f docker-compose.yml -f docker-compose.local.yml up -d
cd tdrs-backend && docker-compose -f docker-compose.yml -f docker-compose.local.yml up -d 
```

1. Open http://localhost:3000/ and sign in.
1. Proceed with functional tests as described herein.
1. Test steps should be captured in the demo GIF(s) and/or screenshots below.
> *Demo GIF(s) and screenshots for testing procedure*

## Deliverables
_More details on how deliverables herein are assessed included [here](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
+ [ ] [**_insert ACs here_**]
+ [ ] **`lfrohlich`** and/or **`adpennington`**  confirmed that ACs are met.

### [Deliverable 2: Tested Code](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [ ] If this PR introduces backend code changes, are they meaningfully tested?
  + [ ] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [ ] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [ ] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [ ] Are backend code style checks passing on CircleCI?
+ [ ] Are frontend code style checks passing on CircleCI?
+ [ ] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?


### [Deliverable 5: Deployed](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [ ] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [ ] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [ ] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [ ] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](../docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

Commit:
saving state real quick

andrew-jameson committed:
Merge branch 'develop' into feat/1613-minimal-retouch

Commit:
finishing merge with latest

Commit:
Missed old test script

Commit:
Added new test, more cleanup

Commit:
Updating unit tests in DFS, preparing for 1610

Commit:
Merging in Jan's 1610 code for parserError useful-ness

Commit:
Revert "Merging in Jan's 1610 code for parserError useful-ness"

This reverts commit c5796da69d0e9a6d356057550378d536e2be5f8b.

Commit:
update to test to use dfs fixture

Commit:
saving state before new 1610 merge

Commit:
Merged develop and 1610 into 1613 latest

Commit:
Resolving merge conflicts with 1610.

Commit:
Linting changes and comparing to 1610

Commit:
Some unit test linting but inherited 1610 issues

Commit:
Re-ordering job to run tests vs lint first.

Commit:
Updates to linting and unit tests.

Commit:
Merge branch 'feature/1610-parser-error-generator' into feat/1613-minimal-retouch

Commit:
Fixing linting.

andrew-jameson commented on tdrs-backend/tdpservice/users/test/test_permissions.py:
```diff
@@ -111,6 +111,9 @@ def test_ofa_system_admin_permissions(ofa_system_admin):
         'parsers.add_parsererror',
         'parsers.change_parsererror',
         'parsers.view_parsererror',
+        'parsers.add_datafilesummary',
+        'parsers.view_datafilesummary',
+        'parsers.change_datafilesummary',
```
This is needed to pass a permissions unit test.

andrew-jameson commented on .circleci/build-and-test/jobs.yml:
```diff
@@ -5,14 +5,14 @@
       - checkout
       - docker-compose-check
       - docker-compose-up-backend
-      - run:
-          name: Execute Python Linting Test
-          command: cd tdrs-backend; docker-compose run --rm web bash -c "flake8 ."
```
Pushing this to later for better Dev Exp.

andrew-jameson commented on tdrs-backend/setup.cfg:
```diff
@@ -34,7 +34,7 @@ exclude =
     */settings/common.py/,
 # D203 conflicts with D211, which is the more conventional of the two
-extend-ignore =  E226,E302,E41,D203
+extend-ignore =  E226,E302,E41,D203,E501
```
```suggestion
extend-ignore =  E226,E302,E41,D203
```

andrew-jameson committed:
Update tdrs-backend/setup.cfg

andrew-jameson commented on tdrs-backend/tdpservice/parsers/models.py:
```diff
@@ -67,3 +67,71 @@ def __str__(self):
+
+    datafile = models.ForeignKey(DataFile, on_delete=models.CASCADE)
+
+    #  eventually needs a breakdown of cases (accepted, rejected, total) per month
+    #  elif qtr2 jan-mar
+    #  elif qtr3 apr-jun
+    #  elif qtr4 jul-sept
```
To be added in second PR.

andrew-jameson commented on tdrs-backend/tdpservice/parsers/models.py:
```diff
@@ -67,3 +67,71 @@ def __str__(self):
+        # to set rejected, we would need to have raised an exception during (pre)parsing
+        # else if there are errors, we can set to accepted with errors
+        # else we can set to accepted (default)
+        if errors == {}:  # This feels better than running len() on `errors`...but is it a dict vs list?
```
Assumption here that we have a dict and not just the list of parserError objects. Could be convinced if we feel it might be worth refactoring other areas to be more functional but this feels good enough.

andrew-jameson commented on tdrs-backend/tdpservice/parsers/models.py:
```diff
@@ -67,3 +67,71 @@ def __str__(self):
+    """Check for pre-parsing errors."""
+    for key in errors.keys():  # keys are 'header', 'trailer', 'document'
+        for error in errors[key]:
+            if error.category == "1":
```
TODO: utilize Jan's new `choices` both here and in the parserError factory.

jtimpe replied:
```python
if error.error_category == ParserErrorCategoryChoices.PRE_CHECK
```
should do it. however, `trailer` errors are reported as `PRE_CHECK` errors, even though the rest of the file should complete parsing. we may want to discard or silence them before setting the status.

andrew-jameson replied:
TODO: comments for expected structure for errors since keys might also include line_number

andrew-jameson replied:
Resolved in 1613.2.

andrew-jameson commented on tdrs-backend/tdpservice/parsers/test/factories.py:
```diff
@@ -15,7 +70,7 @@ class Meta:
     item_number = 1
     field_name = "test field name"
-    category = 1
+    category = "1"
```
TODO: update to use new `choices`

andrew-jameson replied:
Hmm, after playing with the new `ParserErrorCategoryChoices`, they were left flexible such that these still work. It looks more convoluted to try using it formally. Let me know what you think on which direction to take this.

jtimpe replied:
the implementation should look similar to this

```python
from tdpservice.parsers.models import ParserErrorCategoryChoices

# ...
  error_category=ParserErrorCategoryChoices.PRE_CHECK
#...
```

i'm fine with either, especially since the category number is pretty well known/used

raftmsohani replied:
I like the new choices model, it is more readable

andrew-jameson commented on tdrs-backend/tdpservice/parsers/util.py:
```diff
@@ -13,6 +12,15 @@ def value_is_empty(value, length):
     return value is None or value in empty_values
 
 
+error_types = {
+    1: 'File pre-check',
+    2: 'Record value invalid',
+    3: 'Record value consistency',
+    4: 'Case consistency',
+    5: 'Section consistency',
+    6: 'Historical consistency'
+}
+
```
I think this was prototype for the new parserError category `choices`. Should be removed I think?

jtimpe replied:
i've updated the 1113-2 and 1610 base branches to address this as well as the other linting errors.

andrew-jameson commented on tdrs-backend/tdpservice/parsers/util.py:
```diff
@@ -122,6 +129,12 @@ def get_all_fields(self):
         """Get all fields from the schema."""
         return self.fields
 
+    def get_field(self, name):
+        """Get a field from the schema by name."""
+        for field in self.fields:
+            if field.name == name:
+                return field
+
```
I was prototyping this for fetching the startIndex given a name for a logging message. Could be useful in the future, might be worth deleting it as it is unused.

jtimpe commented on tdrs-backend/tdpservice/parsers/models.py:
```diff
@@ -67,3 +67,71 @@ def __str__(self):
+        elif errors:
+            self.status = self.Status.ACCEPTED_WITH_ERRORS
+
+        return self.status
```
should this method also call `.save()`? it's currently handled in `parse.py` on line 126 after `set_status` is called, i'm just curious on convention what's preferred.

raftmsohani replied:
I am always against calling .save in a child method. Prefer to do the save in parse vs model.method

jtimpe replied:
in that case, see my other comment https://github.com/raft-tech/TANF-app/pull/2485/files#r1182567408

andrew-jameson replied:
I believe this is resolved but I can't open that comment link @jtimpe . Please resolve chain if you feel it is good or feel free to ping me or comment here for further clarification.

jtimpe commented on tdrs-backend/tdpservice/parsers/test/test_summary.py:
```diff
@@ -0,0 +1,63 @@
+@pytest.mark.django_db
+def test_dfs_model(dfs):
+    """Test that the model is created and populated correctly."""
+    dfs = dfs
```
are the `dfs = dfs` lines needed?

raftmsohani replied:
pytest should teardown the fixtures after each test and recreate them for the new method test, but I am also curious why we should create another ref?

andrew-jameson replied:
It was a hold-over from before I was using a fixture and instead created a new object with direct call `DataFileSummary.create()`. Yes, these can be removed.

jtimpe commented on tdrs-backend/tdpservice/parsers/test/test_parse.py:
```diff
@@ -26,13 +26,11 @@ def create_test_datafile(filename, stt_user, stt, section='Active Case Data'):
     """Fixture for small_correct_file."""
     return create_test_datafile('small_correct_file', stt_user, stt)
 
-
```
can we add additional `assert`s to the tests in this file to reflect the expected `DatafileSummary.Status`? then i feel the `test_summary.py` tests would not be needed.

andrew-jameson replied:
PM'd you on this, yeah, I end up killing `test_summary.py` in the second PR and we add asserts to the unit tests there.

jtimpe commented on tdrs-backend/tdpservice/parsers/test/test_summary.py:
```diff
@@ -0,0 +1,63 @@
+    test_datafile.save()
+
+    dfs.set_status(parse.parse_datafile(test_datafile))
+    assert dfs.status == DataFileSummary.Status.REJECTED
```
this test is failing for me locally 
```bash
django.db.utils.IntegrityError: duplicate key value violates unique constraint "stts_region_pkey"
```

andrew-jameson replied:
I was seeing this error as well but maybe different test, wiping volume didn't help but setting `transaction=True` in the mark resolved it.

jtimpe commented on tdrs-backend/tdpservice/parsers/test/test_summary.py:
```diff
@@ -0,0 +1,63 @@
+def test_dfs_set_status(dfs):
+    """Test that the status is set correctly."""
+    dfs = dfs
+    assert dfs.status == DataFileSummary.Status.PENDING
```
this test is failing for me locally
```bash
AssertionError: assert <Status.ACCEPTED: 'Accepted'> == <Status.PENDING: 'Pending'>
```

andrew-jameson replied:
Is this still a problem?

jtimpe commented on tdrs-backend/tdpservice/parsers/parse.py:
```diff
@@ -118,9 +117,14 @@ def parse_datafile_lines(datafile, program_type, section):
 
+    summary = DataFileSummary(datafile=datafile)
+    summary.set_status(errors)
+    summary.save()
```
just some food for thought/discussion. my preference for these types of actions is to use django's object manager
```python
summary = DataFileSummary.objects.create(
    datafile=datafile,
    error_category=get_summary_status(errors)
)
```
where `get_summary_status` is a function similar to `DataFileSummary.set_status` except it returns the status instead of updating the object.

this way creating and saving the object is all one step. additionally, `summary.set_status` is inherently impure, having a separate function that gives you an error status to set up the object with, without changing the state of the object itself, is cleaner and leaves less room for misinterpretation.

andrew-jameson replied:
TODO

raftmsohani commented on tdrs-backend/tdpservice/parsers/migrations/0002_alter_parsererror_error_type.py:
```diff
@@ -15,4 +15,13 @@ class Migration(migrations.Migration):
+            name='DataFileSummary',
+            fields=[
+                ('id', models.BigAutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
+                ('status', models.CharField(choices=[('Pending', 'Pending'), ('Accepted', 'Accepted'), ('Accepted with Errors', 'Accepted With Errors'), ('Rejected', 'Rejected')], default='Pending', max_length=50)),
```
Nit: since the choices are limited maybe we can reduce the max_length to 32 or even less

raftmsohani reviewed (commented):
my local test is also failing, might have to nuke the DB

jtimpe committed:
add adminUrl to deployment cypress overrides

Commit:
updates per PR.

Commit:
Excluding trailers for rejection

Commit:
Merge branch 'feature/1610-parser-error-generator' into feat/1613-minimal-retouch

Commit:
VSCode merge resolution is garbage.

Commit:
Fixing precheck for not implemented types

andrew-jameson committed:
Adding "beta" banners to relevant error report sections (#2522)

* Update views.py

* Update views.py

* Update SubmissionHistory.jsx

* Update SubmissionHistory.test.js

* Apply suggestions from code review

Co-authored-by: Miles Reiter <mreiter@goraft.tech>

* lint fixes

---------

Co-authored-by: Miles Reiter <mreiter@goraft.tech>
Co-authored-by: Alex P <63075587+ADPennington@users.noreply.github.com>
Co-authored-by: andrew-jameson <ajameson@teamraft.com>

andrew-jameson committed:
Merge branch 'feature/1610-parser-error-generator' into feat/1613-minimal-retouch

Commit:
Updating to error-handle not implemented schema types

Smithh-Co committed:
Create sprint-73-summary.md (#2565)

andrew-jameson committed:
hotfix for large file sizes (#2542)

* hotfix for large file sizes

* apply timeouts/req limits to dev

* filter identity pages from scan

* IGNORE sql injection

---------

Co-authored-by: Jan Timpe <jtimpe@goraft.tech>
Co-authored-by: mo sohani <msohani@goraft.tech>
Co-authored-by: Alex P <63075587+ADPennington@users.noreply.github.com>

jtimpe committed:
updating validation error language

jtimpe committed:
accidentally included coding challenge

jtimpe committed:
rm comments

elipe17 committed:
2550 deactivation email link (#2557)

* - updated nginx buildpack

* - specifying different nginx version

* - Updating changelog

* - added script to update certain apps in cf
- added workflow for each environment in circi

* - fixed base config

* - fixing jobs

* - Updated based on feedback in OH

* - Updating defaults

* - Removing defaults

* - Fixing indent

* - Adding params to config

* test

* test

* - updating work dir

* - Adding checkout

* - adding cf check

* - logging into cf

* - update cf check to install required binary

* - removing unnecessary switch

* - Forcing plugin installation

* - test installing plugin from script also

* - Adding url to email

* - test code for sandbox

* - using my email

* Revert "Merge branch 'update-cf-os' into 2551-deactivation-email-link"

This reverts commit e963b9df48dd1f72ca0c5b192c979bac11851d11, reversing
changes made to cc9cf81e9d76c42f51ffd5e102f6027d3eb5e645.

* Revert "- using my email"

This reverts commit cc9cf81e9d76c42f51ffd5e102f6027d3eb5e645.

* Revert "- test code for sandbox"

This reverts commit 06037747197d17ed8e63b086fcfcf048ecb50dc4.

---------

Co-authored-by: Alex P <63075587+ADPennington@users.noreply.github.com>
Co-authored-by: Andrew <84722778+andrew-jameson@users.noreply.github.com>

lfrohlich committed:
Update README.md (#2577)

Add ATO

Co-authored-by: Andrew <84722778+andrew-jameson@users.noreply.github.com>

reitermb committed:
Create 2023, Spring - Testing CSV & Excel-based error reports.md

reitermb committed:
Update README.md

andrew-jameson committed:
Updating deliverable links (#2584)

elipe17 committed:
User viewset not returning/duplicating users (#2573)

* - Fixed issue not allowing pagination to work locally with nginx
- Added ordering to user field to fix duplicates issue

* - fix lint error

* - Removing ID check since we cannot guarantee that the uuid that is generated per test run will be lexigraphically consistent

---------

Co-authored-by: Alex P <63075587+ADPennington@users.noreply.github.com>
Co-authored-by: Andrew <84722778+andrew-jameson@users.noreply.github.com>

elipe17 committed:
Update cf os (#2523)

* - updated nginx buildpack

* - specifying different nginx version

* - Updating changelog

* - added script to update certain apps in cf
- added workflow for each environment in circi

* - fixed base config

* - fixing jobs

* - Updated based on feedback in OH

* - Updating defaults

* - Removing defaults

* - Fixing indent

* - Adding params to config

* test

* test

* - updating work dir

* - Adding checkout

* - adding cf check

* - logging into cf

* - update cf check to install required binary

* - removing unnecessary switch

* - Forcing plugin installation

* - test installing plugin from script also

* - Adding new dependencies

* - adding package

* - fixing broken install

* - fixing libs

* - using correct command

* - gettign correct version of libc

* - trying to upgrade libs

* - testing

* - Updated README and script

* Revert "- Updated README and script"

This reverts commit 92697b3e53d1fd87b8d3e7995abb9093aa26e307.

* - Removed unnecessary circi stuff
- Removed script
- Updated docs to callout updating secondary apps

* - Correct spelling error

---------

Co-authored-by: Andrew <84722778+andrew-jameson@users.noreply.github.com>

elipe17 committed:
Item Number Mismatch (#2578)

* - Updated schemas and models to reflect correct item numbers of fields

* - Revert migration

* - Updated header/trailer item numbers

* - Fixed item numbers off by one errors

---------

Co-authored-by: Andrew <84722778+andrew-jameson@users.noreply.github.com>

George-Hudson committed:
pipeline filtering (#2538)

* pipeline changes that filter based on paths and branches. circle ci tracks specified branches in order to keep current functionality on HHS side.

* updated syntax to be in line with build-all.yml

* removed comma

* WIP build flow docs

* added Architecture Decision Record for the change to pipeline workflows

* corrected file type of doc to .md

---------

Co-authored-by: George Hudson <ghudson@teamraft.com>
Co-authored-by: Andrew <84722778+andrew-jameson@users.noreply.github.com>

George-Hudson committed:
Hotfix Devops/2457 path filtering for documentation (#2597)

* pipeline changes that filter based on paths and branches. circle ci tracks specified branches in order to keep current functionality on HHS side.

* updated syntax to be in line with build-all.yml

* removed comma

* WIP build flow docs

* added Architecture Decision Record for the change to pipeline workflows

* corrected file type of doc to .md

* build and test all on PRs even for documentation

---------

Co-authored-by: George Hudson <ghudson@teamraft.com>

Smithh-Co committed:
Create sprint-74-summary.md (#2596)

Co-authored-by: Andrew <84722778+andrew-jameson@users.noreply.github.com>

raftmsohani committed:
added URL filters (#2580)

* added URL filters

* allow github to trigger owasp and label deploys (#2601)

Co-authored-by: George Hudson <ghudson@teamraft.com>

---------

Co-authored-by: Andrew <84722778+andrew-jameson@users.noreply.github.com>
Co-authored-by: George Hudson <georgehudson78@gmail.com>
Co-authored-by: George Hudson <ghudson@teamraft.com>

Smithh-Co committed:
Create sprint-75-summary.md (#2608)

Smithh-Co committed:
Create sprint-76-summary.md (#2609)

Co-authored-by: Andrew <84722778+andrew-jameson@users.noreply.github.com>

elipe17 committed:
- Fixed failing test

elipe17 committed:
- Resolved failing tests

elipe17 committed:
- Corrected merge thrash

elipe17 committed:
- Using randbits to generate pk to get around confilcting sequence pks

elipe17 committed:
Revert "- Using randbits to generate pk to get around confilcting sequence pks"

This reverts commit ac9b0659a62f64c4114c41faf0baa659a92be07c.

elipe17 committed:
- Updating region in fixture instead of factory
- letting django handle transaction for test

elipe17 committed:
Merge branch 'develop' of https://github.com/raft-tech/TANF-app into feat/1613-minimal-retouch

raftmsohani commented on tdrs-backend/tdpservice/parsers/models.py:
```diff
@@ -66,3 +66,86 @@ def __str__(self):
+    """
+    # Do these queries only once, save result during creation of this model
+    # or do we grab this data during parsing and bubble it up during create call?
+    {
```
This sounds like a signal to a separate model, so whenever there is a new DataFileSummary, then the aggregate function runs based on the Datafile.date and adds the number to that new model.

andrew-jameson replied:
Right now this field is just a placeholder but it gets implemented in our .2 PR. Perhaps this comment still applies there? [Check `def case_aggregates_by_month()` out] and let me know.

jtimpe commented on tdrs-backend/tdpservice/parsers/parse.py:
```diff
@@ -82,6 +110,17 @@ def parse_datafile_lines(datafile, program_type, section):
 
         schema = get_schema(line, section, schema_options)
+        if schema is None:
+            errors[line_number] = [util.generate_parser_error(
```
is this needed? both `parse_datafile_line` and `parse_multi_record_line` return a parser error for an incorrect `RecordType` if the schema can't be found (e.g., line 167)

andrew-jameson replied:
They do but I felt that the `generate_error` function used there lost some key data that I have in my direct call. Looking over `make_generate_parser_error` used in those functions, do we lose information like record/field because they're being set to `None` in the call in `parse_datafile_line`? I'm open to removing this if clause but was just me solving a problem which based on the tests isn't a real problem? I'm not convinced it's needed after not touching it for awhile.

Perhaps I used this because I didn't "get" the nested-call situation with an intermediary nickname of `generate_error` that continues to catch me off.
`parse_datafile_line`-> `generate_error` -> `util.make_generate_parser_error` -> `generate_parser_error`

It's also a bit awkward to read:
```
if schema:
  ...
  return records
  
return [(None, False, [generate_error(...)])])
```


I also changed the error msg which I recall was feedback from Alex we had but that's a minor fix.

jtimpe replied:
i'm open to keeping the `field="Record_Type"` as you have it. you're correct though that the `record` is lost from the error - because it's a preparsing error, no record is created. no schema exists in this case either, which is why the field wasn't included initially. i believe the message was also updated as part of #2504. i'm fine with moving the check to this location to consolidate the two, but i'd recommend the existing ones be removed in that case.

jtimpe commented on tdrs-backend/tdpservice/parsers/models.py:
```diff
@@ -66,3 +66,86 @@ def __str__(self):
         """Return the error message."""
         return self.error_message
+
+class DataFileSummary(models.Model):
```
i think it'd be useful to have a django admin for the summary model

andrew-jameson replied:
Done as part of 1613.2!

jtimpe commented on tdrs-backend/tdpservice/parsers/parse.py:
```diff
@@ -62,6 +61,35 @@ def parse_datafile(datafile):
 
     errors = errors | line_errors
 
+    # errors['summary'] = DataFileSummary.objects.create(
```
am I understanding correctly that a `DataFileSummary` is not yet created when the parser task kicks off?
<img width="276" alt="image" src="https://github.com/raft-tech/TANF-app/assets/111305129/0e41d14d-a73c-4ec9-995f-dfaf7e50dab3">

andrew-jameson replied:
Yes, in the second PR we ended up putting this invocation in `scheduling/parser_task.py`

andrew-jameson commented:
Replied to most PR comments today.

Commit:
Removing extra returns for missing record type.

andrew-jameson committed:
Merge branch 'develop' into feat/1613-minimal-retouch

Commit:
lint fix

andrew-jameson commented on tdrs-backend/tdpservice/parsers/models.py:
```diff
@@ -66,3 +66,86 @@ def __str__(self):
+        default=Status.PENDING,
+    )
+
+    datafile = models.ForeignKey(DataFile, on_delete=models.CASCADE)
```
TODO: update to more restrictive version:
https://github.com/raft-tech/TANF-app/blob/0193d9f4e2d8c4573639f841c064197a4e778b0c/tdrs-backend/tdpservice/parsers/models.py#L87

andrew-jameson replied:
Can't actually condense this into previous migration since ~5 other migrations exist between present state and introducing this from merges into this branch. This will get resolved as part of #2622

ADPennington commented:
@andrew-jameson  can you look into the merge conflict here?

andrew-jameson commented:
Closing as obsolete per #2612 as discussed in retro/office hours 8.2.23.
